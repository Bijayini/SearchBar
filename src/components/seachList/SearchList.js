import React from "react";
import PropTypes from 'prop-types';

import {  isStringPresentInElement } from "../../utils/helper";
import Highlight from "../highlight/Highlight";

import('./searchList.scss');

const getItemText = (items, input) => {
  if(!input){
    return '';
  } else if (items.some(item =>isStringPresentInElement(item, input))) {
    return <p className="suggestion-item"><mark>&ldquo;{input}&rdquo;</mark> found in items</p>;
  }
  return '';
};

const SearchList = ({lists, pointer, activeItem, setPointer, noResultFound, input}) => {
  return (
    <ul role="listbox" className="suggestion-container" dataTestId="searchList">
      {lists.length ? lists.map((list, index) => {
        const item = getItemText(list.items, input);
        return (
          <li
            key={list.id}
            ref={index === pointer ? activeItem : null}
            className={index === pointer ? 'suggestion active' : 'suggestion'}
            role="option"
            id={index === pointer ? 'suggestionActive' : 'suggestion'}
            onMouseEnter={() => setPointer(index)}
            onMouseLeave={() => setPointer(-1)}
          >
            <Highlight highlight={input} text={list.id.toUpperCase()} style={{fontSize: '14px',
              fontWeight: 700}} />
            {list.name && <Highlight highlight={input} text={list.name} style={{ fontSize: '13px',
              fontStyle: 'italic'}}/>}
            {list.items.length && item }
            <Highlight highlight={input} text={`${list.address}, ${list.pincode}`} />
          </li>)}) : <li className="no-suggestion">{noResultFound}</li>
        }
    </ul>
  )
};

SearchList.defaultProps = {
  activeItem:null,
  input:'',
  noResultFound:'No  User Found'
};

SearchList.propTypes = {
  lists: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name:PropTypes.string.isRequired,
      address:PropTypes.string.isRequired,
      pincode:PropTypes.string.isRequired,
      items:PropTypes.arrayOf(PropTypes.string),
    })
  ).isRequired,
  pointer:PropTypes.number.isRequired,
  activeItem:PropTypes.object,
  setPointer:PropTypes.func.isRequired,
  input:PropTypes.string,
  noResultFound:PropTypes.string
};

export default SearchList;
