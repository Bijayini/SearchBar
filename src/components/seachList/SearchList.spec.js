import React from "react";
import { render } from "@testing-library/react";
import SearchList from "./SearchList";

describe("SearchList component", () => {
  const props = {
    text:'some data',
    highlight:'So',
    pointer:0,
    activeItem:{
      current:<li>dfghjk</li>
    },
    setPointer:jest.fn(),
    input:'Jo',
    lists:[
      {
            "id": "123-s2-546",
            'name': "John Jacobs",
            "items": ["bucket", "bottle", 'john'],
            "address": "1st Cross, 9th Main, abc Apartment",
            "pincode": "5xx012"
      }
    ]
  };

  it("should render SearchList when input is present", () => {
    const {baseElement} = render(<SearchList {...props} />);

    expect(baseElement).toMatchSnapshot();
  });

  it("should not have items when input is empty", () => {
    const newProps = {
      ...props,
      input: ''
    };
    const {baseElement} = render(<SearchList {...newProps} />);

    expect(baseElement.querySelector('.suggestion-item')).not.toBeInTheDocument();
  });

  it("should not have items when items are not having string", () => {
    const newProps = {
      ...props,
      input: 'Ja'
    };
    const {baseElement} = render(<SearchList {...newProps} />);

    expect(baseElement.querySelector('.suggestion-item')).not.toBeInTheDocument();
  });

});
