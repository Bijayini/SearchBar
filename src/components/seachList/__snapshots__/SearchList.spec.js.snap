// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchList component should render SearchList when input is present 1`] = `
<body>
  <div>
    <ul
      class="suggestion-container"
      datatestid="searchList"
      role="listbox"
    >
      <li
        class="suggestion active"
        id="suggestionActive"
        role="option"
      >
        <p
          style="font-size: 14px; font-weight: 700;"
        >
          <span>
            123-S2-546
          </span>
        </p>
        <p
          style="font-size: 13px; font-style: italic;"
        >
          <mark>
            Jo
          </mark>
          <span>
            hn Jacobs
          </span>
        </p>
        <p
          class="suggestion-item"
        >
          <mark>
            “
            Jo
            ”
          </mark>
           found in items
        </p>
        <p>
          <span>
            1st Cross, 9th Main, abc Apartment, 5xx012
          </span>
        </p>
      </li>
    </ul>
  </div>
</body>
`;
